tests/lib/modem_trace/trace_medium_uart/src/trace_medium_uart_test.c

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <unity.h>
+#include <stdbool.h>
+#include <zephyr/kernel.h>
+#include <string.h>
+
+#include "trace_medium.h"
+#include "mock_nrfx_uarte.h"
+#include "mock_pinctrl.h"
+
+extern int unity_main(void);
+
+static const nrfx_uarte_t *p_uarte_inst_in_use;
+/* Variable to store the event_handler registered by the modem_trace module.*/
+static nrfx_uarte_event_handler_t uarte_callback;
+
+/* Suite teardown shall finalize with mandatory call to generic_suiteTearDown. */
+extern int generic_suiteTearDown(int num_failures);
+
+void nrfx_isr(const void *irq_handler)
+{
+	/* Declared only for pleasing linker. Never expected to be called. */
+	TEST_ASSERT(false);
+}
+
+void setUp(void)
+{
+	mock_nrfx_uarte_Init();
+	mock_pinctrl_Init();
+}
+
+void tearDown(void)
+{
+	mock_nrfx_uarte_Verify();
+	mock_pinctrl_Verify();
+
+	p_uarte_inst_in_use = NULL;
+	uarte_callback = NULL;
+}
+
+static void uart_tx_done_simulate(const uint8_t *const data, size_t len)
+{
+	nrfx_uarte_event_t uarte_event;
+
+	uarte_event.type = NRFX_UARTE_EVT_TX_DONE;
+	uarte_event.data.rxtx.bytes = len;
+	uarte_event.data.rxtx.p_data = (uint8_t *)data;
+
+	uarte_callback(&uarte_event, NULL);
+}
+
+static void uart_tx_error_simulate(void)
+{
+	nrfx_uarte_event_t uarte_event;
+
+	uarte_event.type = NRFX_UARTE_EVT_ERROR;
+
+	uarte_callback(&uarte_event, NULL);
+}
+
+int test_suiteTearDown(int num_failures)
+{
+	return generic_suiteTearDown(num_failures);
+}
+
+static nrfx_err_t nrfx_uarte_init_callback(nrfx_uarte_t const *p_instance,
+					   nrfx_uarte_config_t const *p_config,
+					   nrfx_uarte_event_handler_t event_handler,
+					   int no_of_calls)
+{
+	TEST_ASSERT_NOT_EQUAL(NULL, p_config);
+	TEST_ASSERT_EQUAL(true, p_config->skip_gpio_cfg);
+	TEST_ASSERT_EQUAL(true, p_config->skip_psel_cfg);
+	TEST_ASSERT_EQUAL(NRF_UARTE_HWFC_DISABLED, p_config->hal_cfg.hwfc);
+	TEST_ASSERT_EQUAL(NRF_UARTE_PARITY_EXCLUDED, p_config->hal_cfg.parity);
+	TEST_ASSERT_EQUAL(NRF_UARTE_BAUDRATE_1000000, p_config->baudrate);
+	TEST_ASSERT_EQUAL(1, p_config->interrupt_priority);
+	TEST_ASSERT_EQUAL(NULL, p_config->p_context);
+
+	TEST_ASSERT_NOT_EQUAL(NULL, p_instance);
+	TEST_ASSERT_EQUAL(NRFX_UARTE1_INST_IDX, p_instance->drv_inst_idx);
+	p_uarte_inst_in_use = p_instance;
+
+	TEST_ASSERT_NOT_EQUAL(NULL, event_handler);
+
+	uarte_callback = event_handler;
+
+	return NRFX_SUCCESS;
+}
+
+void initialize_trace_medium_uart_success(void)
```

Comment:
Either have this static or move the content to `test_trace_medium_uart_init_success` and call this instead(?)